### ISSUE TITLE
Increase robustness of test suite

### ISSUE THREAD

**Issue by rennergade** [Lind-Project/lind_project#17]
Our current test suite, through Travis, is minimal, and doesn't give many interesting results besides that the project builds. We should integrate more robust testing into Travis.

**Comment by rennergade** [comment]
Neel has added #25 which helps standardize Travis and build the tests we have. We should be adding additional tests as we go along.

**Comment by rennergade** [comment]
I've been thinking about this a bit more and it definitely would be helpful to round this out sooner than later. At this point we're really just testing the most basic of cases, and really only testing if the program fails catastrophically or not.

Creating a testsuite that we can run with a simple `make test` during the development process, that's also ran by Travis would give us the best coverage. We should check expected outputs as well as failures.

**Comment by rennergade** [comment]
I spoke with @neelchauhan and @Yash151 at the meeting today as well as @jesings online, and wanted to jot this down for reference.

@jesings is working on a test suite which runs as a shell script. The script does the following.

1. The script has two lists of file names, separated by whether the output of the files is expected to be deterministic or not.
2. Compiles all files from both lists with nacl-gcc
3. Copies those files to `lindfs`
4. Runs the deterministic tests, and compares their output to per-determined output (there's some parsing involved to remove headers, but this will diff pure program output)
5. Triggers a Python script to run unit tests on the non-det programs. Tests will most likely have to be added here per program.

Once that framework is done, @neelchauhan and @yash are going to help configure that up with proper outputs and unit tests. @neelchauhan should also configure travis to run this test suite, and we should add this as an option from the top level Makefile as `make tests`

I've added a few more tests that I've been using the last few months to the `test-upgrade` branch. Please include those. In the meantime let's brainstorm and add some more tests to look at syscalls were not currently covering.

**Comment by rennergade** [comment]
Jonathan has added the testing script to the `test_suite` branch. It resides in `lind_project/tests/test_cases`

It takes input as `./suite.sh (file listing non-det tests).txt -d (file listing det tests).txt`

Any programs in the non-deterministic list will need a corresponding python file to test its output. For example `forkandopen.c` has a corresponding `forkandopen.py`. This takes the lind and native outputs as arguments. This usually entails parsing the output and doing some asserts.

I've provisioned the Makefile so this can be run with `make test`.

**Comment by rennergade** [comment]
@mzhang15 I've assigned you to this issue. As I mentioned on our intro call, you're going to be in charge of test suite integration for the first several weeks. Feel free to start on this once you feel comfortable after finishing the architecture/wiki issue.

**Comment by rennergade** [comment]
Since it's related, I've added @mzhang15  to this issue as well: https://github.com/Lind-Project/lind_project/issues/56

**Comment by rennergade** [comment]
This issue has existed for awhile!

Thankfully due a lot of work by @jesings and a bit by me, we have a working suite for several cases. This is now hooked to `make` and some changes were added to the Dockerfile and travis for automation.

I'll need to find another time to add these updates that's convenient for everyone.

I'll close this issue after that, and make a separate issue called "add test cases"

**Comment by rennergade** [comment]
This has now been merged. issue closed!
